File: client/app/technical-indicators/stochastic.service.ts
import { OHLC } from './ohlc'
import { Stochastic } from './stochastic'

// http://investexcel.net/how-to-calculate-the-stochastic-oscillator/
export class StochasticService {
    PARAM_1: number = 14;
    PARAM_2: number = 5; // 3


    calculate(ohlcData: OHLC[]): Stochastic[] {
        //debugger;
        let entries = ohlcData.map(x => new Stochastic(x.date, x.open, x.high, x.low, x.close));
        entries.reduce((previousValue, currentValue, currentIndex, array) => {
            let index = currentIndex + 1;

            
            if (index >= this.PARAM_1) {
                // Highest high & Lowest low
                let subarray = array.slice(index - this.PARAM_1, index); 
                currentValue.highestHigh = Math.max(...subarray.map(x => x.high));
                currentValue.lowestLow = Math.min(...subarray.map(x => x.low));

                // %K
                currentValue.percentK = (currentValue.close -  currentValue.lowestLow) / (currentValue.highestHigh - currentValue.lowestLow) * 100;
            }

            // %D
            if (index >= this.PARAM_1 + this.PARAM_2 - 1) {
                let sum = 0;
                for (let i = this.PARAM_2 - 1; i >= 0; i--) {
                    sum += array[index - i - 1].percentK;
                }
                currentValue.percentD = sum / this.PARAM_2;  
            }

            return currentValue;
        }, new Stochastic(new Date(1900), 0, 0, 0, 0));

        return entries;
    }
}

File: server/RuleOne/src/InvSys.RuleOne.Core/Services/ThreeTools/StochasticService.cs
using System.Collections.Generic;
using InvSys.RuleOne.Core.Models.ThreeTools;
using InvSys.StockQuotes.Api.Client.Proxy.Models;
using System.Linq;

namespace InvSys.RuleOne.Core.Services.ThreeTools
{
    public class StochasticService : IStochasticService
    {
        public ICollection<StochasticData> Calculate(ICollection<HistoricalQuote> prices, int param1, int param2)
        {
            var entries = prices.Select(p => new StochasticData(p)).ToList();
            entries.Insert(0, new StochasticData());
            for (var i = 1; i < entries.Count; i++)
            {
                var previousValue = entries[i - 1];
                var currentValue = entries[i];
                var index = i + 1;

                if (index >= param1)
                {
                    // Highest high & Lowest low
                    var subentries = entries.GetRange(index - param1, param1);
                    currentValue.HighestHigh = subentries.Max(x => x.High);
                    currentValue.LowestLow = subentries.Min(x => x.Low);

                    // %K
                    currentValue.PercentK = (currentValue.Close - currentValue.LowestLow) / (currentValue.HighestHigh - currentValue.LowestLow) * 100;
                }

                // %D
                if (index >= param1 + param2 - 1)
                {
                    var sum = 0m;
                    for (var j = param2 - 1; j >= 0; j--)
                    {
                        sum += entries[index - j - 1].PercentK;
                    }
                    currentValue.PercentD = sum / param2;
                }
            }

            return entries;
        }
    }
}
